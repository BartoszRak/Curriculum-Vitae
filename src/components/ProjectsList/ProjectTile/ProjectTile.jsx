import React, { useState } from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import {
  Typography,
  Card,
  CardContent,
  CardActions,
  IconButton,
  Button,
} from '@material-ui/core'
import { Favorite } from '@material-ui/icons'

import { useFavoriteProjectsStorage } from '~hooks'

import useStyle from './ProjectTile.style'

export function ProjectTile({
  classes: overridingClasses,
  project: {
    description, title, url, tags = [], isFavorite,
  },
}) {
  const classes = { ...useStyle(), ...overridingClasses }
  const { toggleProject } = useFavoriteProjectsStorage()
  const [favorite, setFavorite] = useState(isFavorite)

  return (
    <div className={classes.wrapper}>
      <Card className={classes.root}>
        <CardContent>
          <Typography className={classes.title} variant="h4">
            {title}
          </Typography>
          <Typography className={classes.description} variant="body1">
            {description}
          </Typography>
          <div className={classes.tags}>
            {tags.map(tag => (
              <Typography className={classes.tag} key={tag.name} variant="h6">
                {tag.name}
              </Typography>
            ))}
          </div>
        </CardContent>
        <CardActions
          classes={{
            root: classNames(classes.actions, {
              [classes.favoriteActions]: favorite,
            }),
          }}
        >
          <div>
            {Boolean(url) && (
              <a
                aria-label="Open github repository of project"
                className={classes.href}
                href={url}
                rel="noopener noreferrer"
                target="_blank"
              >
                <Button className={classes.hrefButton} color="primary">
                  Details
                </Button>
              </a>
            )}
          </div>
          {Boolean(isFavorite !== undefined) && (
            <IconButton
              className={classNames({ [classes.favoriteIconButton]: favorite })}
              data-testid="project-tile-icon-button"
              onClick={() => {
                setFavorite(!favorite)
                toggleProject(url)
              }}
            >
              <Favorite className={classes.icon} />
            </IconButton>
          )}
        </CardActions>
      </Card>
    </div>
  )
}

ProjectTile.propTypes = {
  classes: PropTypes.object,
  project: PropTypes.shape({
    description: PropTypes.string,
    isFavorite: PropTypes.bool,
    tags: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string,
        url: PropTypes.string,
      })
    ),
    title: PropTypes.string.isRequired,
    url: PropTypes.string,
  }),
}

export default ProjectTile
