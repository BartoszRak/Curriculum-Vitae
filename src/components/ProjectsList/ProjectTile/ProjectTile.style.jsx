import { makeStyles } from '@material-ui/core'

export default makeStyles(({ spacing, palette }) => ({
  actions: {
    borderBottom: '4px solid rgba(0, 0, 0, 0)',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  description: {
    fontWeight: 'normal',
    marginTop: spacing(1),
  },
  favoriteActions: {
    borderBottom: `4px solid ${palette.error.main}`,
  },
  favoriteIconButton: {
    color: palette.error.main,
  },
  href: {
    textDecoration: 'none',
  },
  hrefButton: {
    borderRadius: spacing(1),
    textTransform: 'uppercase',
  },
  icon: {
    height: '1rem',
    width: '1rem',
  },
  root: {
    height: 'auto',
    textDecoration: 'none',
    width: '100%',
  },
  tag: {
    background: palette.primary.main,
    borderRadius: spacing(2),
    color: palette.primary.contrastText,
    fontWeight: 'normal',
    marginBottom: spacing(1),
    marginRight: spacing(2),
    padding: spacing(0, 1.5),
  },
  tags: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginTop: spacing(2),
  },
  title: {
    borderLeft: `4px solid ${palette.primary.main}`,
    fontWeight: 'normal',
    paddingLeft: spacing(1.5),
  },
  wrapper: {
    breakInside: 'avoid',
    columnBreakInside: 'avoid',
    display: 'table',
    height: 'auto',
    paddingBottom: spacing(4),
    paddingTop: 1,
    pageBreakInside: 'avoid',
    textDecoration: 'none',
    width: '100%',
  },
}))
