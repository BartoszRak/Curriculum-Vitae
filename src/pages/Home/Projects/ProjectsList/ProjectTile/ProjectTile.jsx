import React from 'react'
import PropTypes from 'prop-types'
import { withStyles, Paper, Typography } from '@material-ui/core'

import style from './ProjectTile.style'

export function ProjectTile({
  classes, description, name, url, image,
}) {
  return (
    <a className={classes.root} href={url} rel="noopener noreferrer" target="_blank">
      <Paper
        classes={{
          root: classes.paper,
        }}
      >
        <div
          className={classes.background}
          style={{
            backgroundImage: `url(${image})`,
          }}
        />
        <div className={classes.contentWrapper}>
          <Typography className={classes.title} variant="h1">
            {name}
          </Typography>
          <Typography className={classes.description} variant="body1">
            {description}
          </Typography>
        </div>
      </Paper>
    </a>
  )
}

ProjectTile.propTypes = {
  classes: PropTypes.object,
  description: PropTypes.string,
  image: PropTypes.string,
  name: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}

ProjectTile.defaultProps = {
  description: '',
}

export default withStyles(style)(ProjectTile)
