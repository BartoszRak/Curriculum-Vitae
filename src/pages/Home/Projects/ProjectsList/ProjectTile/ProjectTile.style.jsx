import { makeStyles } from '@material-ui/core'

export default makeStyles(({ spacing, palette }) => ({
  actions: {
    borderBottom: '4px solid rgba(0, 0, 0, 0)',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  description: {
    fontWeight: 'normal',
    marginTop: spacing(1),
  },
  favoriteActions: {
    borderBottom: `4px solid ${palette.error.main}`,
  },
  favoriteIconButton: {
    color: palette.error.main,
  },
  href: {
    textDecoration: 'none',
  },
  hrefButton: {
    textTransform: 'uppercase',
  },
  icon: {
    height: '1rem',
    width: '1rem',
  },
  root: {
    breakInside: 'avoid-column',
    columnBreakInside: 'avoid',
    display: 'table',
    height: 'auto',
    marginBottom: spacing(3),
    pageBreakInside: 'avoid',
    textDecoration: 'none',
    width: '100%',
  },
  tag: {
    background: palette.primary.main,
    borderRadius: spacing(2),
    color: palette.primary.contrastText,
    fontWeight: 'normal',
    marginBottom: spacing(1),
    marginRight: spacing(2),
    padding: spacing(0, 1.5),
  },
  tags: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginTop: spacing(2),
  },
  title: {
    borderLeft: `4px solid ${palette.primary.main}`,
    fontWeight: 'normal',
    paddingLeft: spacing(1.5),
  },
}))
