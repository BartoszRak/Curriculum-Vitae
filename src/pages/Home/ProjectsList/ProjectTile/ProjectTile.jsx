import React from 'react'
import PropTypes from 'prop-types'
import { withStyles, Paper, Typography } from '@material-ui/core'

import style from './ProjectTile.style'

export function ProjectTile({
  classes,
  description,
  name,
  url,
  image,
}) {
  return (
    <a href={url} className={classes.linkWrapper} target="_blank">
      <Paper className={classes.root}>
        <div
          className={classes.background}
          style={{
            backgroundImage: `url(${image})`,
          }}
        />
        <div className={classes.contentWrapper}>
          <Typography className={classes.title} variant="h1">
            {name}
          </Typography>
          <Typography className={classes.description} variant="body1">
            {description}
          </Typography>
        </div>
      </Paper>
    </a>
  )
}

ProjectTile.propTypes = {
  classes: PropTypes.object,
  description: PropTypes.string,
  image: PropTypes.string,
  name: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}

ProjectTile.defaultProps = {
  description: '',
}

export default withStyles(style)(ProjectTile)
