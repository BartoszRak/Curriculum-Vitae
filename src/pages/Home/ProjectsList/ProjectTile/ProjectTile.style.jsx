export default ({ spacing, palette }) => ({
  background: {
    height: '100%',
    left: 0,
    opacity: 0.1,
    position: 'absolute',
    top: 0,
    width: '100%',
  },
  root: {
    border: `1px solid ${palette.secondary.main}`,
    height: '100%',
    padding: spacing.unit * 2,
    position: 'relative',
  },
  title: {
    fontSize: '1.2rem',
  },
})
