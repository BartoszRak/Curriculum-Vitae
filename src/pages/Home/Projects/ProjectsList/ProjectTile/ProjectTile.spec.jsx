import React from 'react'
import { render, cleanup } from '@testing-library/react'

import { ProjectTile } from './ProjectTile'

jest.unmock('@material-ui/core')
jest.unmock('@material-ui/icons')

let project

describe('component: ProjectTile', () => {
  afterEach(cleanup)

  beforeEach(() => {
    project = {
      description: 'Test description',
      isFavorite: false,
      tags: [{
        name: 'Test tag name1',
        url: 'http://testtagurl1.com',
      }, {
        name: 'Test tag name2',
        url: 'http://testtagurl2.com',
      }],
      title: 'Test title',
      url: 'http://testurl.com',
    }
  })

  describe('rendering', () => {
    test('match snapshot ', () => {
      const wrapper = render(<ProjectTile project={project} />)

      expect(wrapper).toMatchSnapshot()
    })

    test('render title and description ', () => {
      const { getByText } = render(<ProjectTile project={project} />)

      const titleEl = getByText(project.title)
      const descriptionEl = getByText(project.description)

      expect(titleEl).toBeVisible()
      expect(descriptionEl).toBeVisible()
    })

    test('render valid url ', () => {
      const { getByText } = render(<ProjectTile project={project} />)

      const hrefEl = getByText('Details').parentNode.parentNode

      expect(hrefEl).toHaveAttribute('href', project.url)
    })

    test('render active icon when project is favorite', () => {
      project.isFavorite = true
      const { getByTestId } = render(<ProjectTile project={project} />)

      const iconButtonEl = getByTestId('project-tile-icon-button')
      expect(iconButtonEl.className).toContain('favoriteIconButton')
    })

    test('render tags', () => {
      project.isFavorite = true
      const { getByText } = render(<ProjectTile project={project} />)

      project.tags.forEach(tag => {
        const tagEl = getByText(tag.name)
        expect(tagEl).toBeVisible()
      })
    })
  })
})
