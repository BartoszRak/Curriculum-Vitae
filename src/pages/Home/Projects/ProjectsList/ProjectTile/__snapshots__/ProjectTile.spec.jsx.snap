// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`global: ProjectTile should render as expected 1`] = `
<a
  href="http://testurl.com"
  rel="noopener noreferrer"
  target="_blank"
>
  <PaperMock
    classes={
      Object {
        "root": undefined,
      }
    }
  >
    <div
      style={
        Object {
          "backgroundImage": "url(http://testImage.com/url)",
        }
      }
    />
    <div>
      <TypographyMock
        variant="h1"
      >
        Example title
      </TypographyMock>
      <TypographyMock
        variant="body1"
      >
        Some example description
      </TypographyMock>
    </div>
  </PaperMock>
</a>
`;
